Fix DeepSeek-V2 expert-parallelism failure due to indexing error

skavulya opened the pull request:
# What does this PR do?
Fixes indexing error during multi-card inference with expert parallelism on DeepSeek-V2. The indexing issue causes the following failure:

`python ../gaudi_spawn.py--world_size=2 run_generation.py --model_name_or_path deepseek-ai/DeepSeek-V2-Lite --use_kv_cache --max_new_tokens 100 --batch_size 1 --bf16 --use_hpu_graphs --parallel_strategy "ep" --prompt "DeepSpeed is a machine learning framework"`

Stack trace:

Warming up iteration 1/3                                                                                                             /usr/local/lib/python3.10/dist-packages/transformers/generation/configuration_utils.py:601: UserWarning: do_sample is set to False. However, temperature is set to 0.3 -- this flag is only used in sample-based generation modes. You should set do_sample=True or unset temperature.                                                                                                               warnings.warn(                                                                                                                     /usr/local/lib/python3.10/dist-packages/transformers/generation/configuration_utils.py:606: UserWarning: do_sample is set to False. However, top_p is set to 0.95 -- this flag is only used in sample-based generation modes. You should set do_sample=True or unset top_p.                                                                                                                          warnings.warn(                                                                                                                     Setting pad_token_id to eos_token_id:None for open-end generation.                                                               [rank0]: Traceback (most recent call last):                                                                                          [rank0]:   File "/root/optimum-habana/examples/text-generation/run_generation.py", line 801, in <module>                             [rank0]:     main()                                                                                                                  [rank0]:   File "/root/optimum-habana/examples/text-generation/run_generation.py", line 563, in main                                 [rank0]:     generate(None, args.reduce_recompile)                                                                                   [rank0]:   File "/root/optimum-habana/examples/text-generation/run_generation.py", line 534, in generate                             [rank0]:     outputs = model.generate(                                                                                               [rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context                  [rank0]:     return func(*args, **kwargs)                                                                                            [rank0]:   File "/root/optimum-habana/optimum/habana/transformers/generation/utils.py", line 1477, in generate                       [rank0]:     result = self._sample(                                                                                                  [rank0]:   File "/root/optimum-habana/optimum/habana/transformers/generation/utils.py", line 2458, in _sample                        [rank0]:     outputs = self(                                                                                                         [rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl               [rank0]:     return self._call_impl(*args, **kwargs)                                                                                 [rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1750, in _call_impl                       [rank0]:     return forward_call(*args, **kwargs)                                                                                    [rank0]:   File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/graphs.py", line 745, in forward                [rank0]:     return wrapped_hpugraph_forward(                                                                                        [rank0]:   File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/graphs.py", line 610, in wrapped_hpugraph_forward                                                                                                                                    [rank0]:     outputs = orig_fwd(*args, **kwargs) 

[rank0]:   File "/root/optimum-habana/optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py", line 1918, in forward
[rank0]:     outputs = self.model(
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1847, in _call_impl
[rank0]:     return inner()
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1793, in inner
[rank0]:     result = forward_call(*args, **kwargs)
[rank0]:   File "/root/optimum-habana/optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py", line 1714, in forward
[rank0]:     layer_outputs = decoder_layer(
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1847, in _call_impl
[rank0]:     return inner()
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1793, in inner
[rank0]:     result = forward_call(*args, **kwargs)
[rank0]:   File "/root/optimum-habana/optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py", line 1411, in forward
[rank0]:     hidden_states = self.mlp(hidden_states)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1847, in _call_impl
[rank0]:     return inner()
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1793, in inner
[rank0]:     result = forward_call(*args, **kwargs)
[rank0]:   File "/root/optimum-habana/optimum/habana/transformers/models/deepseek_v2/modeling_deepseek_v2.py", line 700, in forward
[rank0]:     htcore.mark_step()
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/utils/internal.py", line 36, in lazy_wrapper
[rank0]:     func(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/step_closure.py", line 71, in mark_step
[rank0]:     htcore._mark_step(device_str, sync)
[rank0]: RuntimeError: synNodeCreateWithId failed for node: moe_bf16 with synStatus 26 [Generic failure]. .
<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

skavulya commented:
@IT-Forrest @gyou2021 Please review

gyou2021 commented:
> @IT-Forrest @gyou2021 Please review

LGTM. 
It is also useful for inference on a single card when self.expert_slice = math.ceil(config.n_routed_experts / SLICE_MAX_EXPERT) > config.n_routed_experts / SLICE_MAX_EXPERT.

IT-Forrest commented:
> > @IT-Forrest @gyou2021 Please review
> 
> LGTM. It is also useful for inference on a single card when self.expert_slice = math.ceil(config.n_routed_experts / SLICE_MAX_EXPERT) > config.n_routed_experts / SLICE_MAX_EXPERT.

This change seems OK on DeepSeekV2. Yet two comments
- why do we need `GenerationMixin` in `class DeepseekV2ForCausalLM(DeepseekV2PreTrainedModel, GenerationMixin):`?
- BTW, for DeepSeekV3, the change might hit index exceed err. E.g. the config.n_routed_experts = 256, and self.expert_per_tank = config.n_routed_experts = 80, then self.expert_slice = 4, and self.expert_chunk = 64. Then `self.experts[idx * self.expert_chunk + i + expert_offset]` would hit index exceed err.

Please invite the author of DeepSeek-V2 to take a further review of the PR change. Thanks

skavulya commented:
@IT-Forrest I added the `GenerationMixin` to get rid of a warning about calling `prepare_inputs_for_generation` without extending `GenerationMixin`

I'll also run a test on `DeepSeekV2-Lite` with `SLICE_MAX_EXPERT` set so that it mimics the DeepSeek-V3 scenario test that you highlighted to make sure we can handle that edge case correctly.

skavulya commented:
@IT-Forrest @yao-matrix @gyou2021 I fixed the edge case that @IT-Forrest where the index can exceed the max.  

I also addressed another edge case which occurs where `n_routed_experts` is not divisible by number of ranks. In this scenerio, we were skipping some experts. For example, we have 256 experts on 3 ranks, then experts_per_rank=85 per the original v3 code. With SLICE_MAX_EXPERT=80, we get expert_slice=2. With the earlier version of the PR, we set `expert_chunk = experts_per_rank // self.expert_slice`  i.e., 42. This would result in us using 84 experts per rank instead of 85.

By using `expert_chunk = math.ceil(experts_per_rank / self.expert_slice)`, we get expert_chunk of 43. The first chunk will have 43 experts, and the second 42 due to this bound check:
 `experts_max = min((experts_min + self.expert_chunk), (self.ep_rank + 1) * self.experts_per_rank)`

The behavior is unchanged if the experts are divisible by ranks.

IT-Forrest commented:
> @IT-Forrest @yao-matrix @gyou2021 @regisss I fixed the edge case that @IT-Forrest where the index can exceed the max.
> 
> I also addressed another edge case which occurs where `n_routed_experts` is not divisible by number of ranks. In this scenerio, we were skipping some experts. For example, we have 256 experts on 3 ranks, then experts_per_rank=85 per the original v3 code. With SLICE_MAX_EXPERT=80, we get expert_slice=2. With the earlier version of the PR, we set `expert_chunk = experts_per_rank // self.expert_slice` i.e., 42. This would result in us using 84 experts per rank instead of 85.
> 
> By using `expert_chunk = math.ceil(experts_per_rank / self.expert_slice)`, we get expert_chunk of 43. The first chunk will have 43 experts, and the second 42 due to this bound check: `experts_max = min((experts_min + self.expert_chunk), (self.ep_rank + 1) * self.experts_per_rank)`
> 
> The behavior is unchanged if the experts are divisible by ranks.

Thanks for the update @skavulya, I was sick thus late for the response. I can confirm that the EP fix could work without the fix for [TP failure](https://github.com/habana-internal/deepspeed-fork/pull/91). Look forward to merge to the main branch
